pkg/kubelet/cm/dra/manager.go

```diff
@@ -108,48 +108,73 @@ func (m *ManagerImpl) PrepareResources(pod *v1.Pod) error {
 			continue
 		}
 
-		claimInfo := m.cache.get(*claimName, pod.Namespace)
-		if claimInfo == nil {
-			// claim does not exist in cache, create new claimInfo object
-			// to be processed later.
-			claimInfo = newClaimInfoFromResourceClaim(resourceClaim)
-		}
-
-		// We delay checkpointing of this change until this call
-		// returns successfully. It is OK to do this because we
-		// will only return successfully from this call if the
-		// checkpoint has succeeded. That means if the kubelet is
-		// ever restarted before this checkpoint succeeds, the pod
-		// whose resources are being prepared would never have
-		// started, so it's OK (actually correct) to not include it
-		// in the cache.
-		claimInfo.addPodReference(pod.UID)
-
-		if claimInfo.prepared {
-			// Already prepared this claim, no need to prepare it again
-			continue
-		}
-
-		// Loop through all plugins and prepare for calling NodePrepareResources.
-		for _, resourceHandle := range claimInfo.ResourceHandles {
-			// If no DriverName is provided in the resourceHandle, we
-			// use the DriverName from the status
-			pluginName := resourceHandle.DriverName
-			if pluginName == "" {
-				pluginName = resourceClaim.Status.DriverName
+		// Add a defer to make sure we remove references to this pod in the
+		// claimInfo cache in cases where this function returns an error.
+		defer func(claim *resourceapi.ResourceClaim) {
+			if rerr != nil {
+				m.cache.Lock()
+				claimInfo, exists := m.cache.get(claim.Name, claim.Namespace)
+				if exists {
+					claimInfo.deletePodReference(pod.UID)
+				}
+				m.cache.Unlock()
+			}
+		}(resourceClaim)
+
+		// Atomically perform some operations on the claimInfo cache.
+		err = m.cache.withLock(func() error {
+			// Get a reference to the claim info for this claim from the cache.
+			// If there isn't one yet, then add it to the cache.
+			claimInfo, exists := m.cache.get(resourceClaim.Name, resourceClaim.Namespace)
+			if !exists {
+				claimInfo = m.cache.add(newClaimInfoFromClaim(resourceClaim))
 			}
-			claim := &drapb.Claim{
-				Namespace:      resourceClaim.Namespace,
-				Uid:            string(resourceClaim.UID),
-				Name:           resourceClaim.Name,
-				ResourceHandle: resourceHandle.Data,
+
+			// Add a reference to the current pod in the claim info.
+			// We delay checkpointing of this change until this call
+			// returns successfully. It is OK to do this because we
+			// will only return successfully from this call if the
+			// checkpoint has succeeded. That means if the kubelet is
+			// ever restarted before this checkpoint succeeds, the pod
+			// whose resources are being prepared would never have
+			// started, so it's OK (actually correct) to not include it
+			// in the cache.
+			claimInfo.addPodReference(pod.UID)
```

Comment:
Would it be better if we create new claimInfo after `NodePrepareResources` succeeds (preferably for all drivers)? This would help to reduce :) chances for a data race. Adding pod reference can be also done only if claimInfo is already prepared.

Reply:
See https://github.com/kubernetes/kubernetes/pull/124323#discussion_r1578918305. We need to record that NodePrepareResources may have been called, otherwise we can loose track of that and not call the corresponding NodeUnprepareResources.

Reply:
I will do another pass on this tomorrow. The `defer()` was added hastily and was an attempt to avoid needing to either:
1. Introduce a reconcile loop to ensure that unprepare gets called on deleted pods that are still in the podReferences list
2. hold the lock across the calls to prepare/unprepare on the plugin

I really don't want to lock across the plugin callouts, so It seems we may need to introduce a reconcile loop afterall (unless I can come up with something better when I get back to this).